### PR TITLE
Revives old Mafia PR adding security, a mind bogglingly large amount of refactors, some documentation, and other niceities

### PR DESCRIPTION
**Pull request by tralezab**
revival of #53456 from last year, but with some changed design thoughts on moffia. for example, not removing thoughtfeeder, but instead removing assistant.

### New Roles:
Warden: Can go on alert once per game, killing everyone who visits. Careful that you don't teamkill with this!
Security Officer: Protects other players like a medical doctor, but when attacked they will attack the attacker and die instead of the person they're protecting
Head of Security: Can choose to attack and kill someone during the night. If they are innocent, they will die the next day.

### Things to make game nicer to play:
The game will speed up once there are only half the players who joined, halfing all the times in phases. Should allow games that are "set" to resolve quicker

The game no longer adds assistants

Game 

### Refactor stuff for coders:

Renames can_preform_action signal to on_visit, makes things much more readable
Roleblocking is a flag
Lawyer uses it (duh)

all prerequisites for doing an action are now in a proc that checks for visiting and all that
ALL ACTION STUFF IS NO LONGER DONE COMPLETELY DIFFERENTLY FOR EACH ACTION WOOO NO MORE INSANOFLEX CODE

## Why It's Good For The Game

Mafia is a beautiful shitty mess and my love was sparked by some games that randomly started

## Changelog
:cl:
add: MAFIA: Three new security roles have been added for townies!
tweak: MAFIA: speeds up when half the people are dead, thank god
balance: MAFIA: Assistants are no longer in the game. Too boring of a role when mafia rarely starts
fix: MAFIA: Lawyers didn't roleblock like half the roles in the game?!? now they do.
/:cl:

### CODE REVIEW

**Comment by BadSS13Player** [comment]
>balance: Assistants are no longer in the game. Too boring of a role when

WHEN WHAT???

Also, you didn't add MAFIA: to the last two entries.

**Comment by davethwave** [comment]
Getting rid of assistants should be in a separate PR. Even then assistants are for letting people do their own thing without having any responsibility from a job as well as a job for new people to get used to their controls. If assistant is boring then you aren't doing anything with the role. Open a shop, make a space base. I dunno. But I would be very surprised if this was merged with a job removal as part of the package.

**Comment by optimumtact** [comment]
7 day blocks for both of you for not actually reading the PR before clutching your purse.

He's talking about in the mafia minigame not the job on the station

File: code/datums/achievements/mafia_achievements.dm [hunk 1]
```diff
@@ -95,3 +113,17 @@
 	desc = "Managed to get more than 12 votes when put up on trial, jesus christ."
 	database_id = MAFIA_MEDAL_HATED
 	icon = "hated"
+
+/*
```
Comment (optimumtact) [inline code/datums/achievements/mafia_achievements.dm:117]: your PR is not ready while these are here

File: code/modules/mafia/controller.dm [hunk 2]
```diff
@@ -785,58 +814,59 @@
...
+	var/supports_left = 2
+
 	var/mafiareg_left = 2
 	var/mafiaspe_left = 1
-	var/killing_role = prob(50)
-	var/disruptors = killing_role ? 1 : 2 //still required to calculate overflow
-	var/overflow_left = MAFIA_MAX_PLAYER_COUNT - (invests_left + protects_left + miscs_left + mafiareg_left + mafiaspe_left + killing_role + disruptors)
+
+	var/neutral_killing_role = prob(50)
+	// 1-2 disruptors to equal up to 12, depending on if there was a killing role
```
Comment (Jared-Fogle) [inline code/modules/mafia/controller.dm:831]: Move this comment above the line it's talking about.

File: code/modules/mafia/roles.dm [hunk 3]
```diff
@@ -10,19 +10,19 @@
...
-	///how many votes submitted when you vote.
+	var/role_flags = NONE
+	///how many votes submitted when you vote. used in voting, but not victory
 	var/vote_power = 1
-	var/detect_immune = FALSE
-	var/revealed = FALSE
-	var/datum/outfit/revealed_outfit = /datum/outfit/mafia/assistant //the assistants need a special path to call out they were in fact assistant, everything else can just use job equipment
+	///how many votes your role COULD count for, now or later. used in checking victory
+	var/vote_potential = 1
+	var/datum/outfit/revealed_outfit = /datum/outfit/mafia/assistant //what they get equipped with when they are revealed
```
Comment (Jared-Fogle) [inline code/modules/mafia/roles.dm:18]: Make this a doc comment.

File: code/modules/mafia/roles.dm [hunk 4]
```diff
@@ -10,19 +10,19 @@
 	var/mob/living/carbon/human/body
 	var/obj/effect/landmark/mafia/assigned_landmark
 
-	///how many votes submitted when you vote.
+	var/role_flags = NONE
```
Comment (Jared-Fogle) [inline code/modules/mafia/roles.dm:13]: Document this.

File: code/modules/mafia/_defines.dm [hunk 5]
```diff
@@ -25,6 +27,14 @@
 /// role that upsets the game aka obsessed, usually worse for town than mafia but they can vote against mafia
 #define NEUTRAL_DISRUPT "disrupt"
 
+//role flags (special status of roles like detection immune)
+#define ROLE_UNDETECTABLE	(1<<0) //to all forms of detection, shows themselves as an assistant.
+#define ROLE_CAN_KILL		(1<<1) //has the ability to kill at night and thus, blocks the game from ending with other teams alive.
+#define ROLE_UNIQUE			(1<<2) //can only be one in a randomly generated game
+#define ROLE_REVEALED		(1<<3) //role is public to all other players in the game.
+#define ROLE_VULNERABLE		(1<<4) //can not be defended, protected, or any other form of protection. all kills succeed no matter what.
+#define ROLE_ROLEBLOCKED	(1<<5) //cannot perform any actions that night, preselected actions fail
```
Comment (Jared-Fogle) [inline code/modules/mafia/_defines.dm:36]: Doc comments.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by adamkad1** [comment]
Are we copying town of salem yet? they dont have 'townies' so thats nice

**Review by Twaticus — APPROVED** [review]

**Comment by davethwave** [comment]
Here to apologize. Though you were talking about the families gamemode rather than the maffia minigame.

**Pull request merged**